### PR TITLE
Feature: Get whitepaper via timechain

### PR DESCRIPTION
**Pull request by moneymanolis**
Get whitepaper with your own node from the timechain. Works for pruned nodes as well.

### CODE REVIEW

File: src/cryptoadvance/specter/server_endpoints/controller.py [hunk 1]
```diff
@@ -137,3 +139,46 @@ def get_scantxoutset_status():
         "active": app.specter.info["utxorescan"] is not None,
         "progress": app.specter.info["utxorescan"],
     }
+
+
+@app.route("/bitcoin.pdf")
+@login_required
+def get_whitepaper():
+    if not app.specter.info["pruned"]:
```
Comment (ben-kaufman) [inline src/cryptoadvance/specter/server_endpoints/controller.py:147]: would make sense to first check here if the node is connected and on mainnet so when going via direct URL you could return a proper error message

**Comment by ben-kaufman** [comment]
Looks good, tested and works really well. Just left a small comment.

File: src/cryptoadvance/specter/server_endpoints/controller.py [hunk 2]
```diff
@@ -137,3 +139,54 @@ def get_scantxoutset_status():
...
+                pdf += output["result"]["scriptPubKey"]["hex"][cur : cur + 130]
+                cur += 132
+                pdf += output["result"]["scriptPubKey"]["hex"][cur : cur + 130]
+            pdf += outputs_prun[-1]["result"]["scriptPubKey"]["hex"][4:-4]
+        res = make_response(unhexlify(pdf[16:-16]))
+        res.headers.set("Content-Disposition", "attachment")
+        res.headers.set("Content-Type", "application/pdf")
+        return res
+    else:
+        return render_template(
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/server_endpoints/controller.py:189]: @ben-kaufman here is the error messaging.

File: src/cryptoadvance/specter/server_endpoints/controller.py [hunk 3]
```diff
@@ -137,3 +139,54 @@ def get_scantxoutset_status():
...
+            for output in outputs[1:-2]:
+                cur = 6
+                pdf += output[cur : cur + 130]
+                cur += 132
+                pdf += output[cur : cur + 130]
+                cur += 132
+                pdf += output[cur : cur + 130]
+            pdf += outputs[-2][6:-4]
+        else:
+            outputs_prun = app.specter.rpc.multi(
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/server_endpoints/controller.py:165]: @ben-kaufman and here is the rpc.multi implementation for pruned nodes, much faster!

File: src/cryptoadvance/specter/server_endpoints/controller.py (hunk 1 above)
Comment (moneymanolis) [reply src/cryptoadvance/specter/server_endpoints/controller.py]: Added in this commit https://github.com/cryptoadvance/specter-desktop/pull/905/commits/8844aaf6f089404e10f3e4773b99205dda03c353

**Review by ben-kaufman — APPROVED** [review]
Looks great! Thanks!

Tested on a full archival node, maybe @k9ert or @stepansnigirev could try with a pruned one?

**Comment by ben-kaufman** [comment]
Also tested with a remote pruned node over Tor, works great and fast!

**Comment by djpnewton** [comment]
what is meant by 'timechain'?

**Pull request merged**